src/core/state/requests/index.ts

```diff
@@ -1,26 +1,88 @@
 import { createRainbowStore } from '~/core/state/internal/createRainbowStore';
 
-import { ProviderRequestPayload } from '../../transports/providerRequestTransport';
+import type { ProviderRequestPayload } from '../../transports/providerRequestTransport';
+
+import { getSenderHost, getTabIdString } from './utils';
+
+type PromiseResolver = {
+  resolve: (value: object) => void;
+  reject: (error: Error) => void;
+};
 
 export interface PendingRequestsStore {
   pendingRequests: ProviderRequestPayload[];
-  addPendingRequest: (request: ProviderRequestPayload) => void;
+  connectionPromiseResolvers: Map<string, PromiseResolver[]>;
+  addPendingRequest: (request: ProviderRequestPayload) => boolean;
   removePendingRequest: (id: number) => void;
+  addConnectionResolver: (host: string, resolver: PromiseResolver) => void;
+  resolveConnectionRequests: (
+    host: string,
+    result: unknown,
+    isError?: boolean,
+  ) => void;
 }
 
 export const usePendingRequestStore = createRainbowStore<PendingRequestsStore>(
   (set, get) => ({
     pendingRequests: [],
+    connectionPromiseResolvers: new Map(),
     addPendingRequest: (newRequest) => {
+      // Check for duplicate eth_requestAccounts requests from the same host
+      if (newRequest.method === 'eth_requestAccounts') {
+        const tabId = getTabIdString(newRequest);
+        const senderHost = getSenderHost(newRequest);
+        if (senderHost && tabId) {
+          // Check if there's already a pending connection request for this host
+          const existingConnectionRequest = get().pendingRequests.find(
+            (request) => {
+              if (request.method !== 'eth_requestAccounts') return false;
+              const existingHost = getSenderHost(request);
+              const existingTabId = getTabIdString(request);
+              return existingHost === senderHost && existingTabId === tabId;
+            },
+          );
+
+          if (existingConnectionRequest) {
+            // Don't add duplicate - return false to indicate this is a duplicate
+            return false;
+          }
+        }
+      }
+
       const pendingRequests = get().pendingRequests;
       set({ pendingRequests: pendingRequests.concat([newRequest]) });
+      return true;
     },
     removePendingRequest: (id) => {
       const pendingRequests = get().pendingRequests;
       set({
         pendingRequests: pendingRequests.filter((request) => request.id !== id),
       });
     },
+    addConnectionResolver: (host, resolver) => {
```

Comment:
Can you add some notes about the resolver paradigm used here? Trying to wrap my head around the goal.

Reply:
added some in code comments :) Basically it's needed to allow multiple connect requests to resolve when one dialog gets accepted/declined